File: Leetcode/Medium/19. Remove Nth Node From End of List.cpp
class Solution
{
public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        ListNode *fast = head;
        while (n-- > 0)
        {
            fast = fast->next;
        }
        ListNode *prev = nullptr;
        ListNode *slow = head;
        while (fast != nullptr)
        {
            fast = fast->next;
            prev = slow;
            slow = slow->next;
        }
        if (prev != nullptr)
        {
            prev->next = slow->next;
        }
        else
        {
            head = slow->next;
        }
        return head;
    }
};

File: Leetcode/19. Remove Nth Node From End of List/19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode()
        dummy.next = head
        curr = dummy
        prev = None
        for i in range(n):
            curr = curr.next
        tmp = dummy
        prev = None
        while curr:
            curr = curr.next
            prev = tmp
            tmp = tmp.next
        prev.next = tmp.next
        return dummy.next